library/core/src/fmt/mod.rs

```diff
@@ -2399,23 +2399,68 @@ impl Display for bool {
 impl Debug for str {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result {
         f.write_char('"')?;
-        let mut from = 0;
-        for (i, c) in self.char_indices() {
-            let esc = c.escape_debug_ext(EscapeDebugExtArgs {
-                escape_grapheme_extended: true,
-                escape_single_quote: false,
-                escape_double_quote: true,
-            });
-            // If char needs escaping, flush backlog so far and write, else skip
-            if esc.len() != 1 {
-                f.write_str(&self[from..i])?;
-                for c in esc {
-                    f.write_char(c)?;
+
+        // substring we know is printable
+        let mut printable_range = 0..0;
+
+        // the outer loop here splits the string into ASCII-only, and Unicode-only chunks,
+        // which are then processed separately, to enable a fast path for the ASCII-only chunk.
+        let mut rest = self.as_bytes();
+        while rest.len() > 0 {
+            let mut ascii_bytes: &[u8];
+            let unicode_bytes: &[u8];
+
+            // first, handle an ascii-only prefix
+            let non_ascii_position = rest.iter().position(|&b| b >= 0x80).unwrap_or(rest.len());
+            // SAFETY: the position was derived from an iterator, so is known to be within bounds, and at a char boundary
+            (ascii_bytes, rest) = unsafe { rest.split_at_unchecked(non_ascii_position) };
+
+            fn needs_escape(b: u8) -> bool {
+                b > 0x7E || b < 0x20 || b == b'\\' || b == b'"'
+            }
+            while let Some(escape_position) = ascii_bytes.iter().position(|&b| needs_escape(b)) {
```

Comment:
I wonder if it would be faster to combine this check with the ASCII check, and combine this loop with the Unicode loop below. I.e. find the next character that is not an ASCII printable, then escape it, check if it is easily printable, and then continue the loop.

Reply:
This makes a ton of sense, now I just skip ascii-printable prefixes, and then process a chunk of non-ascii-printable chars one by one.
I might do another iteration on the code that skips the two nested loops and rather runs the outer loop per non-ascii-printable char. Lets see how that turns out perf-wise.

Thus far I am quite surprised by whatever the optimizer is doing, as you can see in the most recent commit.